File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.cc
#include <iostream>
#include <vector>
#include <queue>

#define FAST_IO() \
    ios::sync_with_stdio(false); \
    cin.tie(NULL); \
    cout.tie(NULL);
#define MAX_WEIGHT 200000000
#define ll long long

using namespace std;

int n;

struct node {
    int index;
    ll dist;
    node() {}
    node(int index, ll dist) : index(index), dist(dist) {}
    int operator < (const node other) const {
        return this->dist > other.dist;
    }
};

vector<ll> dijkstra(vector<vector<int> >& graph, int source) {
    vector<ll> weight(n, MAX_WEIGHT);
    priority_queue<node> queue;
    queue.push(node(source, 0));
    weight[source] = 0;

    while (!queue.empty()) {
        node tmp = queue.top();
        queue.pop();
        if (weight[tmp.index] < tmp.dist) continue;
        for (int i = 0; i < n; i++) {
            if (graph[tmp.index][i] == 0) continue;
            ll value = weight[tmp.index] + graph[tmp.index][i];
            if (weight[i] > value) {
                weight[i] = value;
                queue.push(node(i, value));
            }
        }
    }
    return weight;
}

int main() {
// int algorithm() {
    FAST_IO();

    int e;
    cin >> n >> e;

    vector<vector<int> > graph(n, vector<int>(n, 0));
    int u, v, w;
    while (e-- > 0) {
        cin >> u >> v >> w;
        graph[u - 1][v - 1] = w;
        graph[v - 1][u - 1] = w;
    }


    int d1, d2;
    cin >> d1 >> d2;

    vector<ll> fromStart = dijkstra(graph, 0),
        toEnd = dijkstra(graph, n - 1),
        fromD1 = dijkstra(graph, d1 - 1);

     ll path1 = fromStart[d1 - 1] + fromD1[d2 - 1] + toEnd[d2 - 1],
        path2 = fromStart[d2 - 1] + fromD1[d2 - 1] + toEnd[d1 - 1];

    if (min(path1, path2) >= MAX_WEIGHT) cout << -1 << '\n';
    else cout << min(path1, path2) << '\n';

    return 0;
}

File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.java
import java.io.*;
import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
    private static class Node {
        int nextVertex, weight;
        Node link;

        Node(int nextVertex, int weight, Node link) {
            this.nextVertex = nextVertex;
            this.weight = weight;
            this.link = link;
        }
    }
    private static final int INF = 2_000_000_001;
    public static void main(String[] args) throws IOException {
        try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
             BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out))) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken()), E = Integer.parseInt(st.nextToken());
            Node[] graph = new Node[N + 1];
            for (int i = 0; i < E; i++) {
                st = new StringTokenizer(br.readLine());
                int a = Integer.parseInt(st.nextToken()), b = Integer.parseInt(st.nextToken()), c = Integer.parseInt(st.nextToken());
                graph[a] = new Node(b, c, graph[a]);
                graph[b] = new Node(a, c, graph[b]);
            }
            st = new StringTokenizer(br.readLine());
            int v1 = Integer.parseInt(st.nextToken()), v2 = Integer.parseInt(st.nextToken());
            int[] fromStart = dijkstra(1, graph), toEnd = dijkstra(N, graph), fromV = dijkstra(v1, graph);
            long path1 = (long) fromStart[v1] + fromV[v2] + toEnd[v2], path2 = (long) fromStart[v2] + fromV[v2] + toEnd[v1];
            if (Math.min(path1, path2) >= INF) bw.write(String.valueOf(-1));
            else bw.write(String.valueOf(Math.min(path1, path2)));
            bw.flush();
        }
    }
    static int[] dijkstra(int start, Node[] graph) {
        int[] weight = new int[graph.length];
        Arrays.fill(weight, INF);
        PriorityQueue<int[]> priorityQueue = new PriorityQueue<>(Comparator.comparingInt(array -> array[1]));
        priorityQueue.offer(new int[]{start, 0});
        weight[start] = 0;
        while (!priorityQueue.isEmpty()) {
            int[] tmp = priorityQueue.poll();
            if (weight[tmp[0]] < tmp[1]) continue;
            for (Node node = graph[tmp[0]]; node != null; node = node.link) {
                int value = weight[tmp[0]] + node.weight;
                if (weight[node.nextVertex] > value) {
                    weight[node.nextVertex] = value;
                    priorityQueue.offer(new int[]{node.nextVertex, value});
                }
            }
        }
        return weight;
    }
}